lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,625 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/Allocator.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#define DEBUG_TYPE "gct"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Utilities
+//===----------------------------------------------------------------------===//
+
+/// An absolute instance path.
+using InstancePath = ArrayRef<InstanceOp>;
+
+template <typename T>
+T &operator<<(T &os, const InstancePath &path) {
+  os << "$root";
+  for (auto inst : path)
+    os << "." << inst.name();
+  return os;
+}
+
+namespace {
+
+/// A port annotated with a data tap key or mem tap.
+struct AnnotatedPort {
+  unsigned portNum;
+  Annotation anno;
+};
+
+/// An extmodule that has annotated ports.
+struct AnnotatedExtModule {
+  FExtModuleOp extModule;
+  SmallVector<AnnotatedPort, 4> portAnnos;
+  ArrayAttr filteredModuleAnnos; // module annotations without data tap stuff
+};
+
+/// A value annotated to be tapped.
+struct TappedValue {
+  Value value;
+  Annotation anno;
+};
+
+/// A data structure that tracks the instances of modules and can provide
+/// absolute paths to these instances.
+struct InstanceGraph {
+  /// The root circuit.
+  CircuitOp circuitOp;
+  /// The main module in the circuit.
+  Operation *mainModule;
+  /// A mapping from a module to all its instances in the design.
+  DenseMap<Operation *, SmallVector<InstanceOp, 1>> moduleInstances;
+
+  InstanceGraph(CircuitOp circuitOp);
+  ArrayRef<InstancePath> getAbsolutePaths(Operation *op);
+
+private:
+  /// An allocator for individual instance paths and entire path lists.
+  llvm::BumpPtrAllocator allocator;
+
+  /// Cached absolute instance paths.
+  DenseMap<Operation *, ArrayRef<InstancePath>> absolutePathsCache;
+
+  /// Append an instance to a path.
+  InstancePath appendInstance(InstancePath path, InstanceOp inst);
+};
+
+/// Necessary information to wire up a port with tapped data or memory location.
+struct PortWiring {
+  unsigned portNum;
+  ArrayRef<InstancePath> prefices;
+  SmallString<16> suffix;
+};
+
+} // namespace
+
+InstanceGraph::InstanceGraph(CircuitOp circuitOp) : circuitOp(circuitOp) {
+  mainModule = circuitOp.getMainModule();
+
+  // Gather all instances in the circuit.
+  circuitOp.walk([&](Operation *op) {
+    if (auto instOp = dyn_cast<InstanceOp>(op))
+      moduleInstances[instOp.getReferencedModule()].push_back(instOp);
+  });
+
+  LLVM_DEBUG(llvm::dbgs() << "Instance graph:\n");
```

Comment:
Old habit, but I'd throw #ifndef NDEBUG around the debug code since it has loops

Reply:
Good catch, that slipped through the cracks. Done.